Figure out a policy for rank-n types

What do you think the following code does?

```hs
$(singletons [d|
  f :: (forall a. a -> a) -> b -> b
  f _ x = x
  |])
```

To my complete surprise, this code actually _compiles_. I was baffled as to how this could possibly work until I looked at the `-ddump-splices` output, which I've produced below (after some minor cleanup):

```hs
    type family F (f :: a ~> a) (x :: b) :: b where
      F _ x = x
    sF ::
      forall b a (f :: a ~> a) (x :: b).
      Sing f -> Sing x -> Sing (F f x)
    sF _ sX = sX
```

Yikes. Unlike the type of `f`, which uses higher-rank polymorphism, neither the kind of `F` nor the type of `sF` are higher-rank! It seems that `singletons` drops the nested `forall` and, by sheer accident, this allows the generated code to compile.

This is clearly less than ideal, and we should aspire to do better. But what is the right thing to do when `singletons` is presented with rank-_n_ types? I contemplated how `singletons` might be changed so that the code it generates is also rank-_n_, but there are a number of obstacles that stand in the way of this vision:

## 1(a): Defunctionalization symbols (current `singletons`)

If we were to promote the `f` example above, here is a defunctionalization symbol that would need to be generated:

```hs
data FooSym0 :: (forall a. a ~> a) ~> b ~> b
```

But note that `(forall a. a ~> a) ~> b ~> b` is the same thing as `TyFun (forall a. a ~> a) (b ~> b) -> Type`, which is decidedly impredicative. I doubt that we want to require users to enable `ImpredicativeTypes`.

## 1(b): Defunctionalization symbols (alternative encoding of `(~>)`)

In an attempt to work around 1(a), I was reminded of [this alternative way](https://github.com/goldfirere/singletons/issues/320#issue-311443055) to encode `(~>)`:

```hs
type a ~> b = a -> b -> Type
```

I thought this might be worth considering because with this encoding, `(forall a. a ~> a) ~> b ~> b` would become `(forall a. a -> a -> Type) -> (b -> b -> Type) -> Type`, which isn't impredicative anymore! Alas, this doesn't get you very far. If you try to define an instance of `Apply` using this version of `(~>)`:

```hs
type instance forall (x :: forall a. a ~> a).
  Apply FSym0 x = FSym1 x
```

Then GHC is not happy:

```
Foo.hs:33:9: error:
    • Expected kind ‘(a0 ~> a0) ~> (b0 ~> b0)’,
        but ‘FSym0’ has kind ‘(forall a. a ~> a) -> (b0 ~> b0) -> *’
    • In the first argument of ‘Apply’, namely ‘FSym0’
      In the type instance declaration for ‘Apply’
   |
33 |   Apply FSym0 x = FSym1 x
   |         ^^^^^
```

It seems that GHC is reluctant to infer a higher-rank kind for `FSym0` in that spot. I tried to make GHC come to its senses:

```hs
type instance forall (x :: forall a. a ~> a).
  Apply @(forall a. a ~> a) FSym0 x = FSym1 x
```

But that makes GHC even less happy:

```
Foo.hs:33:3: error:
    • Illegal polymorphic type: forall a. a ~> a
    • In the type instance declaration for ‘Apply’
   |
33 |   Apply @(forall a. a ~> a) FSym0 x = FSym1 x
   |   ^^^^^
```

Bah. At this point, I gave up on the idea. (Man, I wish `(~>)` were [a part of GHC itself](https://github.com/ghc-proposals/ghc-proposals/pull/242) so that we didn't have to deal with defunctionalization-related crap...)

## 2: In singled definitions

Even if you ignore the challenges that defunctionalization presents, there's still another hurdle to clear: singling. What should the singled version of this function be?

```hs
g :: (forall a. a -> a) -> b -> b
g h x = h x
```

Here is a lousy first attempt:

```hs
type family G (h :: forall a. a ~> a) (x :: b) :: b where
  forall b (h :: forall a. a ~> a) (x :: b). G h x = h @@ x

sG :: forall b (h :: forall a. a ~> a) (x :: b).
      Sing h -> Sing x -> Sing (G h x)
sG sH sX = applySing sH sX
```

Sadly, `sG` does not typecheck:

```
Foo.hs:36:25: error:
    • Couldn't match type ‘x’ with ‘t0’
      ‘x’ is a rigid type variable bound by
        the type signature for:
          sG :: forall a1 b (h :: forall a2. a2 ~> a2) (x :: b).
                Sing h -> Sing x -> Sing (G h x)
        at Foo.hs:(34,1)-(35,38)
      Expected type: Sing t0
        Actual type: Sing x
    • In the second argument of ‘applySing’, namely ‘sX’
      In the expression: applySing sH sX
      In an equation for ‘sG’: sG sH sX = applySing sH sX
    • Relevant bindings include
        sX :: Sing x (bound at Foo.hs:36:7)
        sH :: Sing h (bound at Foo.hs:36:4)
        sG :: Sing h -> Sing x -> Sing (G h x) (bound at Foo.hs:36:1)
   |
36 | sG sH sX = applySing sH sX
   |                         ^^
```

One way to work around this is to use the following type signature for `sG` instead:

```hs
sG :: forall b (h :: forall a. a ~> a) (x :: b).
      (forall a. Sing (h @a)) -> Sing x -> Sing (G h x)
sG sH sX = applySing sH sX
```

This works, but it doesn't scale well, since this trick can only work for rank-_n_ types where the `forall`s always appear in prenex form. That is to say, if you tried to single this definition:

```hs
j :: (a -> forall b. b -> b) -> a -> c -> c
j k x y = k x y
```

Then you're hosed, since there's no way to instantiate the `b` in `a ~> forall b. b ~> b` with visible kind application alone.

-----

All of these challenges lead me to believe that we should simply make any attempt to promote or single a rank-_n_ type an error. In fact, `singletons` currently implements a half-hearted attempt to do this here:

https://github.com/goldfirere/singletons/blob/27245e24d588a63620ffb7f0276345ac36f8a6ba/src/Data/Singletons/Promote/Type.hs#L27-L29

In practice, that `(DAppT (DAppT DArrowT (DForallT (_:_) _ _)) _)` case is very rarely exercised, since all top-level function types are passed to `promoteUnraveled` first, which splits apart `DArrowT`s and then promotes the individual function argument types. As a consequence, it would catch rank-3 (or above) types, but not rank-2 types.

If we do ban rank-_n_ types, what should become of types with nested `forall`s, such as in the following example?

```hs
$(singletons [d|
  const2 :: forall a. a -> forall b. b -> a
  const2 x _ = x
  |])
```

Again, this code actually compiles today, but only because `singletons` treats it as if you had written `forall a b. a -> b -> a`. Should these be banned alongside other rank-_n_ types, or should we grandfather in these sorts of functions that do anything that is explicitly higher-rank?

--- a/.travis.yml
+++ b/.travis.yml
@@ -111,7 +111,7 @@ install:
     echo "source-repository-package"                            >> cabal.project
     echo "  type:     git"                                      >> cabal.project
     echo "  location: https://github.com/goldfirere/th-desugar" >> cabal.project
-    echo "  tag:      35cdb6c294651c5b55733cb3348f7722300c92a7" >> cabal.project
+    echo "  tag:      3a1eadc51049bfae66fcb7873428fdb06f3040c3" >> cabal.project
     echo ""                                                     >> cabal.project
     echo "package th-desugar"                                   >> cabal.project
     echo "  tests:      False"                                  >> cabal.project
@@ -146,7 +146,7 @@ script:
     echo "source-repository-package"                            >> cabal.project
     echo "  type:     git"                                      >> cabal.project
     echo "  location: https://github.com/goldfirere/th-desugar" >> cabal.project
-    echo "  tag:      35cdb6c294651c5b55733cb3348f7722300c92a7" >> cabal.project
+    echo "  tag:      3a1eadc51049bfae66fcb7873428fdb06f3040c3" >> cabal.project
     echo ""                                                     >> cabal.project
     echo "package th-desugar"                                   >> cabal.project
     echo "  tests:      False"                                  >> cabal.project

--- a/.travis.yml
+++ b/.travis.yml
@@ -4,7 +4,7 @@
 #
 # For more information, see https://github.com/haskell-CI/haskell-ci
 #
-# version: 0.3.20190815
+# version: 0.4
 #
 language: c
 dist: xenial
@@ -111,7 +111,7 @@ install:
     echo "source-repository-package"                            >> cabal.project
     echo "  type:     git"                                      >> cabal.project
     echo "  location: https://github.com/goldfirere/th-desugar" >> cabal.project
-    echo "  tag:      d8d00ea8dd19dac3628435cac7a31ab0f214db9b" >> cabal.project
+    echo "  tag:      35cdb6c294651c5b55733cb3348f7722300c92a7" >> cabal.project
     echo ""                                                     >> cabal.project
     echo "package th-desugar"                                   >> cabal.project
     echo "  tests:      False"                                  >> cabal.project
@@ -146,7 +146,7 @@ script:
     echo "source-repository-package"                            >> cabal.project
     echo "  type:     git"                                      >> cabal.project
     echo "  location: https://github.com/goldfirere/th-desugar" >> cabal.project
-    echo "  tag:      d8d00ea8dd19dac3628435cac7a31ab0f214db9b" >> cabal.project
+    echo "  tag:      35cdb6c294651c5b55733cb3348f7722300c92a7" >> cabal.project
     echo ""                                                     >> cabal.project
     echo "package th-desugar"                                   >> cabal.project
     echo "  tests:      False"                                  >> cabal.project

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,6 +1,14 @@
 Changelog for singletons project
 ================================
 
+next
+----
+* Require GHC 8.10.
+* `singletons` now does a more much thorough job of rejecting higher-rank types
+  during promotion or singling, as they cannot support them. (Previously,
+  `singletons` would sometimes accept them, often changing rank-2 types to
+  rank-1 types incorrectly in the process.)
+
 2.6
 ---
 * Require GHC 8.8.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -5,9 +5,9 @@ next
 ----
 * Require GHC 8.10.
 * `singletons` now does a more much thorough job of rejecting higher-rank types
-  during promotion or singling, as they cannot support them. (Previously,
-  `singletons` would sometimes accept them, often changing rank-2 types to
-  rank-1 types incorrectly in the process.)
+  during promotion or singling, as `singletons` cannot support them.
+  (Previously, `singletons` would sometimes accept them, often changing rank-2
+  types to rank-1 types incorrectly in the process.)
 
 2.6
 ---

--- a/cabal.project
+++ b/cabal.project
@@ -4,4 +4,4 @@ packages: .
 source-repository-package
   type: git
   location: https://github.com/goldfirere/th-desugar
-  tag: 35cdb6c294651c5b55733cb3348f7722300c92a7
+  tag: 3a1eadc51049bfae66fcb7873428fdb06f3040c3

--- a/cabal.project
+++ b/cabal.project
@@ -4,4 +4,4 @@ packages: .
 source-repository-package
   type: git
   location: https://github.com/goldfirere/th-desugar
-  tag: d8d00ea8dd19dac3628435cac7a31ab0f214db9b
+  tag: 35cdb6c294651c5b55733cb3348f7722300c92a7

--- a/singletons.cabal
+++ b/singletons.cabal
@@ -59,7 +59,7 @@ library
                       ghc-boot-th,
                       template-haskell,
                       containers >= 0.5,
-                      th-desugar >= 1.10 && < 1.11,
+                      th-desugar >= 1.11 && < 1.12,
                       pretty,
                       syb >= 0.4,
                       text >= 1.2,

--- a/src/Data/Singletons/CustomStar.hs
+++ b/src/Data/Singletons/CustomStar.hs
@@ -133,7 +133,8 @@ singletonStar names = do
 
         -- demote a kind back to a type, accumulating any unbound parameters
         kindToType :: DsMonad q => [DTypeArg] -> DKind -> QWithAux [Name] q DType
-        kindToType _    (DForallT _ _ _) = fail "Explicit forall encountered in kind"
+        kindToType _    (DForallT _ _ _)    = fail "Explicit forall encountered in kind"
+        kindToType _    (DConstrainedT _ _) = fail "Explicit constraint encountered in kind"
         kindToType args (DAppT f a) = do
           a' <- kindToType [] a
           kindToType (DTANormal a' : args) f

--- a/src/Data/Singletons/Deriving/Show.hs
+++ b/src/Data/Singletons/Deriving/Show.hs
@@ -241,5 +241,5 @@ mk_Show_arg_pat ForShowSing{} arg arg_ty =
 mk_Show_rhs_sig :: ShowMode -> [Name] -> DExp -> DExp
 mk_Show_rhs_sig ForPromotion  _            e = e
 mk_Show_rhs_sig ForShowSing{} arg_ty_names e =
-  e `DSigE` DForallT [] (map (DAppT (DConT showSing'Name) . DVarT) arg_ty_names)
-                        (DConT showSName)
+  e `DSigE` DConstrainedT (map (DAppT (DConT showSing'Name) . DVarT) arg_ty_names)
+                          (DConT showSName)

--- a/src/Data/Singletons/Deriving/Util.hs
+++ b/src/Data/Singletons/Deriving/Util.hs
@@ -137,14 +137,15 @@ functorLikeTraverse var (FT { ft_triv = caseTrivial, ft_var = caseVar
                                       -- error if we encounter one here.
                                   then pure (caseWrongArg, True)
                                   else tyApp
-                             inspect (DForallT _ _ t) = inspect t
-                             inspect (DSigT t _)      = inspect t
-                             inspect (DAppT t _)      = inspect t
-                             inspect (DAppKindT t _)  = inspect t
-                             inspect (DVarT {})       = tyApp
-                             inspect DArrowT          = tyApp
-                             inspect (DLitT {})       = tyApp
-                             inspect DWildCardT       = tyApp
+                             inspect (DForallT _ _ t)    = inspect t
+                             inspect (DConstrainedT _ t) = inspect t
+                             inspect (DSigT t _)         = inspect t
+                             inspect (DAppT t _)         = inspect t
+                             inspect (DAppKindT t _)     = inspect t
+                             inspect (DVarT {})          = tyApp
+                             inspect DArrowT             = tyApp
+                             inspect (DLitT {})          = tyApp
+                             inspect DWildCardT          = tyApp
 
                          in case unfoldDType f of
                               (f_head, _) -> inspect f_head
@@ -162,11 +163,12 @@ functorLikeTraverse var (FT { ft_triv = caseTrivial, ft_var = caseVar
     go (DVarT v)
       | v == var = pure (caseVar, True)
       | otherwise = trivial
-    go (DForallT tvbs _ t) = do
+    go (DForallT _ tvbs t) = do
       (tr, tc) <- go t
       if var `notElem` map extractTvbName tvbs && tc
          then pure (caseForAll tvbs tr, True)
          else trivial
+    go (DConstrainedT _ t) =  go t
     go (DConT {}) = trivial
     go DArrowT    = trivial
     go (DLitT {}) = trivial

--- a/src/Data/Singletons/Names.hs
+++ b/src/Data/Singletons/Names.hs
@@ -320,7 +320,9 @@ splitUnderscores s = case span (== '_') s of
 modifyConNameDType :: (Name -> Name) -> DType -> DType
 modifyConNameDType mod_con_name = go
   where
-    go (DForallT tvbs cxt p) = DForallT tvbs (map go cxt) (go p)
+    go :: DType -> DType
+    go (DForallT fvf tvbs p) = DForallT fvf tvbs (go p)
+    go (DConstrainedT cxt p) = DConstrainedT (map go cxt) (go p)
     go (DAppT     p t)       = DAppT     (go p) t
     go (DAppKindT p k)       = DAppKindT (go p) k
     go (DSigT     p k)       = DSigT     (go p) k

--- a/src/Data/Singletons/Promote/Defun.hs
+++ b/src/Data/Singletons/Promote/Defun.hs
@@ -214,7 +214,7 @@ defunctionalize name m_fixity m_arg_tvbs' m_res_kind' = do
                                          toposortTyVarsOf $     -- (2)(iii)(c)
                                          map tvb_to_type $      -- (2)(iii)(b)
                                          toList $ fvDType tyfun -- (2)(iii)(a)
-                        in (arg_params, Just (DForallT tyfun_tvbs [] tyfun))
+                        in (arg_params, Just (DForallT ForallInvis tyfun_tvbs tyfun))
             app_data_ty = foldTypeTvbs (DConT data_name) m_args
             app_eqn     = DTySynEqn Nothing
                                     (DConT applyName `DAppT` app_data_ty
@@ -249,10 +249,16 @@ defunctionalize name m_fixity m_arg_tvbs' m_res_kind' = do
     eta_expand :: [DTyVarBndr] -> Maybe DKind -> PrM ([DTyVarBndr], Maybe DKind)
     eta_expand m_arg_tvbs Nothing = pure (m_arg_tvbs, Nothing)
     eta_expand m_arg_tvbs (Just res_kind) = do
-        let (_, _, argKs, resultK) = unravel res_kind
-        tvb_names <- replicateM (length argKs) $ qNewName "e"
-        let res_kind_arg_tvbs = zipWith DKindedTV tvb_names argKs
-        pure (m_arg_tvbs ++ res_kind_arg_tvbs, Just resultK)
+        let (arg_ks, result_k) = unravelDType res_kind
+            vis_arg_ks = filterDVisFunArgs arg_ks
+        extra_arg_tvbs <- traverse mk_extra_tvb vis_arg_ks
+        pure (m_arg_tvbs ++ extra_arg_tvbs, Just result_k)
+
+    mk_extra_tvb :: DVisFunArg -> PrM DTyVarBndr
+    mk_extra_tvb vfa =
+      case vfa of
+        DVisFADep tvb -> pure tvb
+        DVisFAAnon k  -> DKindedTV <$> qNewName "e" <*> pure k
 
     map_tvb_kind :: (DKind -> DKind) -> DTyVarBndr -> DTyVarBndr
     map_tvb_kind _ tvb@DPlainTV{}  = tvb

--- a/src/Data/Singletons/Promote/Type.hs
+++ b/src/Data/Singletons/Promote/Type.hs
@@ -7,26 +7,37 @@ This file implements promotion of types into kinds.
 -}
 
 module Data.Singletons.Promote.Type
-  ( promoteType, promoteTypeArg, promoteUnraveled
+  ( promoteType, promoteType_NC
+  , promoteTypeArg_NC, promoteUnraveled
   ) where
 
 import Language.Haskell.TH.Desugar
 import Data.Singletons.Names
+import Data.Singletons.Util
 import Language.Haskell.TH
 
--- the only monadic thing we do here is fail. This allows the function
--- to be used from the Singletons module
+-- Promote a DType to the kind level.
+--
+-- NB: the only monadic thing we do here is fail. This allows the function
+-- to be used from the Singletons module.
 promoteType :: MonadFail m => DType -> m DKind
-promoteType = go []
+promoteType ty = do
+  checkVanillaDType ty
+  promoteType_NC ty
+
+-- Promote a DType to the kind level. This is suffixed with "_NC" because
+-- we do not invoke checkVanillaDType here.
+-- See [Vanilla-type validity checking during promotion].
+promoteType_NC :: MonadFail m => DType -> m DKind
+promoteType_NC = go []
   where
     go :: MonadFail m => [DTypeArg] -> DType -> m DKind
+    go []       (DForallT _fvf _tvbs ty) = go [] ty
     -- We don't need to worry about constraints: they are used to express
     -- static guarantees at runtime. But, because we don't need to do
     -- anything special to keep static guarantees at compile time, we don't
     -- need to promote them.
-    go []       (DForallT _tvbs _cxt ty) = go [] ty
-    go []       (DAppT (DAppT DArrowT (DForallT (_:_) _ _)) _) =
-      fail "Cannot promote types of rank above 1."
+    go []       (DConstrainedT _cxt ty) = go [] ty
     go args     (DAppT t1 t2) = do
       k2 <- go [] t2
       go (DTANormal k2 : args) t1
@@ -60,14 +71,49 @@ promoteType = go []
                             "headed by: " ++ show hd ++ "\n" ++
                             "applied to: " ++ show args
 
-promoteTypeArg :: MonadFail m => DTypeArg -> m DTypeArg
-promoteTypeArg (DTANormal t) = DTANormal <$> promoteType t
-promoteTypeArg ta@(DTyArg _) = pure ta -- Kinds are already promoted
+-- | Promote a DTypeArg to the kind level. This is suffixed with "_NC" because
+-- we do not invoke checkVanillaDType here.
+-- See [Vanilla-type validity checking during promotion].
+promoteTypeArg_NC :: MonadFail m => DTypeArg -> m DTypeArg
+promoteTypeArg_NC (DTANormal t) = DTANormal <$> promoteType_NC t
+promoteTypeArg_NC ta@(DTyArg _) = pure ta -- Kinds are already promoted
 
+-- | Promote a DType to the kind level, splitting it into its argument and
+-- result types in the process.
 promoteUnraveled :: MonadFail m => DType -> m ([DKind], DKind)
 promoteUnraveled ty = do
-  arg_kis <- mapM promoteType arg_tys
-  res_ki  <- promoteType res_ty
+  checkVanillaDType ty
+  arg_kis <- mapM promoteType_NC arg_tys
+  res_ki  <- promoteType_NC res_ty
   return (arg_kis, res_ki)
   where
-    (_, _, arg_tys, res_ty) = unravel ty
+    (_, _, arg_tys, res_ty) = unravelVanillaDType ty
+
+{-
+Note [Vanilla-type validity checking during promotion]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+We only support promoting (and singling) vanilla types, where a vanilla
+function type is a type that:
+
+1. Only uses a @forall@ at the top level, if used at all. That is to say, it
+   does not contain any nested or higher-rank @forall@s.
+
+2. Only uses a context (e.g., @c => ...@) at the top level, if used at all,
+   and only after the top-level @forall@ if one is present. That is to say,
+   it does not contain any nested or higher-rank contexts.
+
+3. Contains no visible dependent quantification.
+
+The checkVanillaDType function checks if a type is vanilla. Note that it is
+crucial to call checkVanillaDType on the /entire/ type. For instance, it would
+be incorrect to call unravelVanillaDType and then check each argument type
+individually, since that loses information about which @forall@s/constraints
+are higher-rank.
+
+We make an effort to avoiding calling checkVanillaDType on the same type twice,
+since checkVanillaDType must traverse the entire type. (It would not be
+incorrect to do so, just wasteful.) For this certain, certain functions are
+suffixed with "_NC" (short for "no checking") to indicate that they do not
+invoke checkVanillaDType. These functions are used on types that have already
+been validity-checked.
+-}

--- a/src/Data/Singletons/Promote/Type.hs
+++ b/src/Data/Singletons/Promote/Type.hs
@@ -32,7 +32,9 @@ promoteType_NC :: MonadFail m => DType -> m DKind
 promoteType_NC = go []
   where
     go :: MonadFail m => [DTypeArg] -> DType -> m DKind
-    go []       (DForallT _fvf _tvbs ty) = go [] ty
+    go []       (DForallT fvf tvbs ty) = do
+      ty' <- go [] ty
+      pure $ DForallT fvf tvbs ty'
     -- We don't need to worry about constraints: they are used to express
     -- static guarantees at runtime. But, because we don't need to do
     -- anything special to keep static guarantees at compile time, we don't

--- a/src/Data/Singletons/Single.hs
+++ b/src/Data/Singletons/Single.hs
@@ -250,9 +250,9 @@ singITyConInstance n
            mk_fun arrow t1 t2 = arrow `DAppT` t1 `DAppT` t2
            matchable_apply_fun   = mk_fun DArrowT                (DVarT k_penult) (DVarT k_last)
            unmatchable_apply_fun = mk_fun (DConT tyFunArrowName) (DVarT k_penult) (DVarT k_last)
-           ctxt = [ DForallT (map DPlainTV as)
-                             (map (DAppT (DConT singIName)) a_tys)
-                             (DConT singIName `DAppT` foldType f_ty a_tys)
+           ctxt = [ DForallT ForallInvis (map DPlainTV as) $
+                    DConstrainedT (map (DAppT (DConT singIName)) a_tys)
+                                  (DConT singIName `DAppT` foldType f_ty a_tys)
                   , DConT equalityName
                       `DAppT` (DConT applyTyConName `DSigT`
                                 mk_fun DArrowT matchable_apply_fun unmatchable_apply_fun)
@@ -410,9 +410,10 @@ singClassD (ClassDecl { cd_cxt  = cls_cxt
                                 -- See #175
                                [ foldApply (promoteValRhs meth_name) tvs `DSigT` res_ki
                                , foldApply prom_dflt tvs ]
-      return $ DForallT tvbs (default_pred : cxt) (ravel args res)
+      return $ DForallT ForallInvis tvbs
+             $ DConstrainedT (default_pred : cxt) (ravel args res)
       where
-        (tvbs, cxt, args, res) = unravel sty
+        (tvbs, cxt, args, res) = unravelVanillaDType sty
         bound_kv_set = Set.fromList bound_kvs
         -- Filter out explicitly bound kind variables. Otherwise, if you had
         -- the following class (#312):
@@ -486,9 +487,9 @@ singInstD (InstDecl { id_cxt = cxt, id_name = inst_name, id_arg_tys = inst_tys
         Nothing -> case mb_s_info of
           -- We don't have an InstanceSig, so we must compute the type to use
           -- in the singled instance ourselves through reification.
-          Just (DVarI _ (DForallT cls_tvbs _cls_pred s_ty) _) -> do
+          Just (DVarI _ (DForallT _ cls_tvbs (DConstrainedT _cls_pred s_ty)) _) -> do
             let subst = mk_subst cls_tvbs
-                (sing_tvbs, ctxt, _args, res_ty) = unravel s_ty
+                (sing_tvbs, ctxt, _args, res_ty) = unravelVanillaDType s_ty
                 m_res_ki = case res_ty of
                   _sing `DAppT` (_prom_func `DSigT` res_ki) -> Just (substKind subst res_ki)
                   _                                         -> Nothing
@@ -500,7 +501,8 @@ singInstD (InstDecl { id_cxt = cxt, id_name = inst_name, id_arg_tys = inst_tys
           _ -> do
             mb_info <- dsReify name
             case mb_info of
-              Just (DVarI _ (DForallT cls_tvbs _cls_pred inner_ty) _) -> do
+              Just (DVarI _ (DForallT _ cls_tvbs
+                                      (DConstrainedT _cls_pred inner_ty)) _) -> do
                 let subst = mk_subst cls_tvbs
                     cls_kvb_names = foldMap (foldMap fvDType . extractTvbKind) cls_tvbs
                     cls_tvb_names = OSet.fromList $ map extractTvbName cls_tvbs
@@ -604,7 +606,8 @@ singTySig defns types bound_kvs name prom_ty =
     mk_sing_ty :: Int -> SgM (DType, [Name])
     mk_sing_ty n = do
       arg_names <- replicateM n (qNewName "arg")
-      return ( DForallT (map DPlainTV arg_names) []
+      return ( DForallT ForallInvis
+                        (map DPlainTV arg_names)
                         (ravel (map (\nm -> singFamily `DAppT` DVarT nm) arg_names)
                                (singFamily `DAppT`
                                     (foldl apply prom_ty (map DVarT arg_names))))

--- a/src/Data/Singletons/Single/Data.hs
+++ b/src/Data/Singletons/Single/Data.hs
@@ -66,7 +66,7 @@ singDataD (DataDecl name tvbs ctors) = do
         DTySynInstD $ DTySynEqn Nothing
                                 (DConT singFamilyName `DAppKindT` k)
                                 (DConT singDataName)
-      kindedSingTy = DForallT (map DPlainTV tvbNames) [] $
+      kindedSingTy = DForallT ForallInvis (map DPlainTV tvbNames) $
                      DArrowT `DAppT` k `DAppT` DConT typeKindName
 
   return $ (DDataD Data [] singDataName [] (Just kindedSingTy) ctors' []) :

--- a/src/Data/Singletons/Single/Type.hs
+++ b/src/Data/Singletons/Single/Type.hs
@@ -31,12 +31,13 @@ singType :: OSet Name      -- the set of bound kind variables in this scope
                 , [DKind]  -- the kinds of the argument types
                 , DKind )  -- the kind of the result type
 singType bound_kvs prom ty = do
-  let (_, cxt, args, res) = unravel ty
+  checkVanillaDType ty
+  let (_, cxt, args, res) = unravelVanillaDType ty
       num_args            = length args
-  cxt' <- mapM singPred cxt
+  cxt' <- mapM singPred_NC cxt
   arg_names <- replicateM num_args (qNewName "t")
-  prom_args <- mapM promoteType args
-  prom_res  <- promoteType res
+  prom_args <- mapM promoteType_NC args
+  prom_res  <- promoteType_NC res
   let args' = map (\n -> singFamily `DAppT` (DVarT n)) arg_names
       res'  = singFamily `DAppT` (foldl apply prom (map DVarT arg_names) `DSigT` prom_res)
       tau   = ravel args' res'
@@ -45,15 +46,30 @@ singType bound_kvs prom ty = do
       kv_names_to_bind = foldMap fvDType (prom_args ++ cxt' ++ [prom_res])
                             OSet.\\ bound_kvs
       kvs_to_bind      = toList kv_names_to_bind
-  let ty' = DForallT (map DPlainTV kvs_to_bind ++ zipWith DKindedTV arg_names prom_args)
-                     cxt' tau
+  let ty' = DForallT ForallInvis
+                     (map DPlainTV kvs_to_bind ++ zipWith DKindedTV arg_names prom_args) $
+            DConstrainedT cxt' tau
   return (ty', num_args, arg_names, cxt, prom_args, prom_res)
 
+-- Single a DPred, checking that it is a vanilla type in the process.
+-- See [Vanilla-type validity checking during promotion]
+-- in Data.Singletons.Promote.Type.
 singPred :: DPred -> SgM DPred
-singPred = singPredRec []
+singPred p = do
+  checkVanillaDType p
+  singPred_NC p
 
+-- Single a DPred. Does not check if the argument is a vanilla type.
+-- See [Vanilla-type validity checking during promotion]
+-- in Data.Singletons.Promote.Type.
+singPred_NC :: DPred -> SgM DPred
+singPred_NC = singPredRec []
+
+-- The workhorse for singPred_NC.
 singPredRec :: [DTypeArg] -> DPred -> SgM DPred
 singPredRec _cxt (DForallT {}) =
+  fail "Singling of quantified constraints not yet supported"
+singPredRec _cxt (DConstrainedT {}) =
   fail "Singling of quantified constraints not yet supported"
 singPredRec ctx (DAppT pr ty) = singPredRec (DTANormal ty : ctx) pr
 singPredRec ctx (DAppKindT pr ki) = singPredRec (DTyArg ki : ctx) pr
@@ -65,7 +81,7 @@ singPredRec ctx (DConT n)
   | n == equalityName
   = fail "Singling of type equality constraints not yet supported"
   | otherwise = do
-    kis <- mapM promoteTypeArg ctx
+    kis <- mapM promoteTypeArg_NC ctx
     let sName = singClassName n
     return $ applyDType (DConT sName) kis
 singPredRec _ctx DWildCardT = return DWildCardT  -- it just might work

--- a/src/Data/Singletons/Util.hs
+++ b/src/Data/Singletons/Util.hs
@@ -10,11 +10,12 @@ Users of the package should not need to consult this file.
 {-# LANGUAGE TypeSynonymInstances, FlexibleInstances, RankNTypes,
              TemplateHaskell, GeneralizedNewtypeDeriving,
              MultiParamTypeClasses, UndecidableInstances, MagicHash,
-             LambdaCase, NoMonomorphismRestriction #-}
+             LambdaCase, NoMonomorphismRestriction, ScopedTypeVariables #-}
 
 module Data.Singletons.Util where
 
 import Prelude hiding ( exp, foldl, concat, mapM, any, pred )
+import Language.Haskell.TH ( pprint )
 import Language.Haskell.TH.Syntax hiding ( lift )
 import Language.Haskell.TH.Desugar
 import Data.Char
@@ -251,10 +252,130 @@ ravel :: [DType] -> DType -> DType
 ravel []    res  = res
 ravel (h:t) res = DAppT (DAppT DArrowT h) (ravel t res)
 
+-- Decompose a vanilla function type into its type variables, its context, its
+-- argument types, and its result type. (See
+-- Note [Vanilla-type validity checking during promotion] in
+-- Data.Singletons.Promote.Type for what "vanilla" means.)
+-- If a non-vanilla construct is encountered while decomposing the function
+-- type, an error is thrown.
+--
+-- This should be contrasted with the 'unravelDType' function from
+-- @th-desugar@, which supports the full gamut of function types. @singletons@
+-- only supports a subset of these types, which is why this function is used
+-- to decompose them instead.
+unravelVanillaDType :: DType -> ([DTyVarBndr], DCxt, [DType], DType)
+unravelVanillaDType ty =
+  let (args1, res)  = unravelDType ty
+      (args2, tvbs) = take_tvbs  args1
+      (args3, ctxt) = take_ctxt args2
+      anons         = take_anons args3
+  in (tvbs, ctxt, anons, res)
+  where
+    take_tvbs :: DFunArgs -> (DFunArgs, [DTyVarBndr])
+    take_tvbs (DFAForalls ForallInvis tvbs args) =
+      let (args', tvbs') = take_tvbs args
+      in (args', tvbs ++ tvbs')
+    take_tvbs (DFAForalls ForallVis _ _) = vdq_err
+    take_tvbs args = (args, [])
+
+    take_ctxt :: DFunArgs -> (DFunArgs, DCxt)
+    take_ctxt (DFACxt ctxt args) =
+      let (args', ctxt') = take_ctxt args
+      in (args', ctxt ++ ctxt')
+    take_ctxt (DFAForalls fvf _ _) =
+      case fvf of
+        ForallInvis -> nested_forall_err
+        ForallVis   -> vdq_err
+    take_ctxt args = (args, [])
+
+    take_anons :: DFunArgs -> [DType]
+    take_anons (DFAAnon anon args) = anon : take_anons args
+    take_anons (DFAForalls fvf _ _) =
+      case fvf of
+        ForallInvis -> nested_forall_err
+        ForallVis   -> vdq_err
+    take_anons (DFACxt _ _) = nested_ctxt_err
+    take_anons DFANil = []
+
+    vdq_err :: a
+    vdq_err =
+      error $ "Unexpected visible dependent quantification in type: " ++ show ty
+
+    nested_forall_err :: a
+    nested_forall_err =
+      error $ "Unexpected nested forall in type: " ++ show ty
+
+    nested_ctxt_err :: a
+    nested_ctxt_err =
+      error $ "Unexpected nested context in type: " ++ show ty
+
+-- Ensures that a 'DType' is a vanilla type. (See
+-- Note [Vanilla-type validity checking during promotion] in
+-- Data.Singletons.Promote.Type for what "vanilla" means.)
+--
+-- The only monadic thing that this function can do is 'fail', which it does
+-- if a non-vanilla construct is encountered.
+checkVanillaDType :: forall m. MonadFail m => DType -> m ()
+checkVanillaDType ty = go_ty True ty
+  where
+    go_ty :: Bool -- Are invisible arguments permitted?
+          -> DType -> m ()
+    go_ty invis_ok typ = do
+      let (args, _) = unravelDType typ
+      go_args invis_ok args
+
+    go_args :: Bool -> DFunArgs -> m ()
+    go_args = go_tvbs
+
+    go_tvbs :: Bool -> DFunArgs -> m ()
+    go_tvbs invis_ok (DFAForalls ForallInvis tvbs args) = do
+      unless invis_ok $ fail_forall "higher-rank"
+      traverse_ (traverse_ (go_ty False) . extractTvbKind) tvbs
+      go_tvbs invis_ok args
+    go_tvbs _ (DFAForalls ForallVis _ _) = fail_vdq
+    go_tvbs invis_ok args = go_ctxt invis_ok args
+
+    go_ctxt :: Bool -> DFunArgs -> m ()
+    go_ctxt invis_ok (DFACxt ctxt args) = do
+      unless invis_ok $ fail_ctxt "higher-rank"
+      traverse_ (go_ty False) ctxt
+      go_ctxt invis_ok args
+    go_ctxt _ (DFAForalls fvf _ _) =
+      case fvf of
+        ForallInvis -> fail_forall "nested"
+        ForallVis   -> fail_vdq
+    go_ctxt _ args = go_anons args
+
+    go_anons :: DFunArgs -> m ()
+    go_anons (DFAAnon anon args) = do
+      go_ty False anon
+      go_anons args
+    go_anons (DFAForalls fvf _ _) =
+      case fvf of
+        ForallInvis -> fail_forall "nested"
+        ForallVis   -> fail_vdq
+    go_anons (DFACxt _ _) = fail_ctxt "nested"
+    go_anons DFANil = pure ()
+
+    failWith :: String -> m ()
+    failWith thing = fail $ unlines
+      [ "`singletons` does not support " ++ thing
+      , "In the type: " ++ pprint (sweeten ty)
+      ]
+
+    fail_forall :: String -> m ()
+    fail_forall sort = failWith $ sort ++ " `forall`s"
+
+    fail_vdq :: m ()
+    fail_vdq = failWith "visible dependent quantification"
+
+    fail_ctxt :: String -> m ()
+    fail_ctxt sort = failWith $ sort ++ " contexts"
+
 -- count the number of arguments in a type
 countArgs :: DType -> Int
-countArgs ty = length args
-  where (_, _, args, _) = unravel ty
+countArgs ty = length $ filterDVisFunArgs args
+  where (args, _) = unravelDType ty
 
 -- changes all TyVars not to be NameU's. Workaround for GHC#11812
 noExactTyVars :: Data a => a -> a
@@ -283,12 +404,13 @@ substKind = substType
 -- substitution, use @substTy@ from "Language.Haskell.TH.Desugar.Subst".
 substType :: Map Name DType -> DType -> DType
 substType subst ty | Map.null subst = ty
-substType subst (DForallT tvbs cxt inner_ty)
-  = DForallT tvbs' cxt' inner_ty'
+substType subst (DForallT fvf tvbs inner_ty)
+  = DForallT fvf tvbs' inner_ty'
   where
     (subst', tvbs') = mapAccumL subst_tvb subst tvbs
-    cxt'            = map (substType subst') cxt
     inner_ty'       = substType subst' inner_ty
+substType subst (DConstrainedT cxt inner_ty) =
+  DConstrainedT (map (substType subst) cxt) (substType subst inner_ty)
 substType subst (DAppT ty1 ty2) = substType subst ty1 `DAppT` substType subst ty2
 substType subst (DAppKindT ty ki) = substType subst ty `DAppKindT` substType subst ki
 substType subst (DSigT ty ki) = substType subst ty `DSigT` substType subst ki

--- a/src/Data/Singletons/Util.hs
+++ b/src/Data/Singletons/Util.hs
@@ -10,7 +10,8 @@ Users of the package should not need to consult this file.
 {-# LANGUAGE TypeSynonymInstances, FlexibleInstances, RankNTypes,
              TemplateHaskell, GeneralizedNewtypeDeriving,
              MultiParamTypeClasses, UndecidableInstances, MagicHash,
-             LambdaCase, NoMonomorphismRestriction, ScopedTypeVariables #-}
+             LambdaCase, NoMonomorphismRestriction, ScopedTypeVariables,
+             FlexibleContexts #-}
 
 module Data.Singletons.Util where
 
@@ -20,8 +21,9 @@ import Language.Haskell.TH.Syntax hiding ( lift )
 import Language.Haskell.TH.Desugar
 import Data.Char
 import Control.Monad hiding ( mapM )
-import Control.Monad.Writer hiding ( mapM )
+import Control.Monad.Except hiding ( mapM )
 import Control.Monad.Reader hiding ( mapM )
+import Control.Monad.Writer hiding ( mapM )
 import qualified Data.Map as Map
 import Data.List.NonEmpty (NonEmpty(..))
 import Data.Map ( Map )
@@ -265,49 +267,9 @@ ravel (h:t) res = DAppT (DAppT DArrowT h) (ravel t res)
 -- to decompose them instead.
 unravelVanillaDType :: DType -> ([DTyVarBndr], DCxt, [DType], DType)
 unravelVanillaDType ty =
-  let (args1, res)  = unravelDType ty
-      (args2, tvbs) = take_tvbs  args1
-      (args3, ctxt) = take_ctxt args2
-      anons         = take_anons args3
-  in (tvbs, ctxt, anons, res)
-  where
-    take_tvbs :: DFunArgs -> (DFunArgs, [DTyVarBndr])
-    take_tvbs (DFAForalls ForallInvis tvbs args) =
-      let (args', tvbs') = take_tvbs args
-      in (args', tvbs ++ tvbs')
-    take_tvbs (DFAForalls ForallVis _ _) = vdq_err
-    take_tvbs args = (args, [])
-
-    take_ctxt :: DFunArgs -> (DFunArgs, DCxt)
-    take_ctxt (DFACxt ctxt args) =
-      let (args', ctxt') = take_ctxt args
-      in (args', ctxt ++ ctxt')
-    take_ctxt (DFAForalls fvf _ _) =
-      case fvf of
-        ForallInvis -> nested_forall_err
-        ForallVis   -> vdq_err
-    take_ctxt args = (args, [])
-
-    take_anons :: DFunArgs -> [DType]
-    take_anons (DFAAnon anon args) = anon : take_anons args
-    take_anons (DFAForalls fvf _ _) =
-      case fvf of
-        ForallInvis -> nested_forall_err
-        ForallVis   -> vdq_err
-    take_anons (DFACxt _ _) = nested_ctxt_err
-    take_anons DFANil = []
-
-    vdq_err :: a
-    vdq_err =
-      error $ "Unexpected visible dependent quantification in type: " ++ show ty
-
-    nested_forall_err :: a
-    nested_forall_err =
-      error $ "Unexpected nested forall in type: " ++ show ty
-
-    nested_ctxt_err :: a
-    nested_ctxt_err =
-      error $ "Unexpected nested context in type: " ++ show ty
+  case unravelVanillaDType_either ty of
+    Left err      -> error err
+    Right payload -> payload
 
 -- Ensures that a 'DType' is a vanilla type. (See
 -- Note [Vanilla-type validity checking during promotion] in
@@ -316,61 +278,90 @@ unravelVanillaDType ty =
 -- The only monadic thing that this function can do is 'fail', which it does
 -- if a non-vanilla construct is encountered.
 checkVanillaDType :: forall m. MonadFail m => DType -> m ()
-checkVanillaDType ty = go_ty True ty
+checkVanillaDType ty =
+  case unravelVanillaDType_either ty of
+    Left err -> fail err
+    Right _  -> pure ()
+
+-- The workhorse that powers unravelVanillaDType and checkVanillaDType.
+-- Returns @Right payload@ upon success, and @Left error_msg@ upon failure.
+unravelVanillaDType_either ::
+  DType -> Either String ([DTyVarBndr], DCxt, [DType], DType)
+unravelVanillaDType_either ty =
+  runIdentity $ flip runReaderT True $ runExceptT $ runUnravelM $ go_ty ty
   where
-    go_ty :: Bool -- Are invisible arguments permitted?
-          -> DType -> m ()
-    go_ty invis_ok typ = do
-      let (args, _) = unravelDType typ
-      go_args invis_ok args
+    go_ty :: DType -> UnravelM ([DTyVarBndr], DCxt, [DType], DType)
+    go_ty typ = do
+      let (args1, res) = unravelDType typ
+      (args2, tvbs) <- take_tvbs  args1
+      (args3, ctxt) <- take_ctxt  args2
+      anons         <- take_anons args3
+      pure (tvbs, ctxt, anons, res)
 
-    go_args :: Bool -> DFunArgs -> m ()
-    go_args = go_tvbs
+    -- Process a type in a higher-order position (e.g., the @forall a. a -> a@ in
+    -- @(forall a. a -> a) -> b -> b@). This is only done to check for the
+    -- presence of higher-rank foralls or constraints, which are not permitted
+    -- in vanilla types.
+    go_higher_order_ty :: DType -> UnravelM ()
+    go_higher_order_ty typ = () <$ local (const False) (go_ty typ)
 
-    go_tvbs :: Bool -> DFunArgs -> m ()
-    go_tvbs invis_ok (DFAForalls ForallInvis tvbs args) = do
-      unless invis_ok $ fail_forall "higher-rank"
-      traverse_ (traverse_ (go_ty False) . extractTvbKind) tvbs
-      go_tvbs invis_ok args
-    go_tvbs _ (DFAForalls ForallVis _ _) = fail_vdq
-    go_tvbs invis_ok args = go_ctxt invis_ok args
+    take_tvbs :: DFunArgs -> UnravelM (DFunArgs, [DTyVarBndr])
+    take_tvbs (DFAForalls ForallInvis tvbs args) = do
+      rank_1 <- ask
+      unless rank_1 $ fail_forall "higher-rank"
+      _ <- traverse_ (traverse_ go_higher_order_ty . extractTvbKind) tvbs
+      (args', tvbs') <- take_tvbs args
+      pure (args', tvbs ++ tvbs')
+    take_tvbs (DFAForalls ForallVis _ _) = fail_vdq
+    take_tvbs args = pure (args, [])
 
-    go_ctxt :: Bool -> DFunArgs -> m ()
-    go_ctxt invis_ok (DFACxt ctxt args) = do
-      unless invis_ok $ fail_ctxt "higher-rank"
-      traverse_ (go_ty False) ctxt
-      go_ctxt invis_ok args
-    go_ctxt _ (DFAForalls fvf _ _) =
+    take_ctxt :: DFunArgs -> UnravelM (DFunArgs, DCxt)
+    take_ctxt (DFACxt ctxt args) = do
+      rank_1 <- ask
+      unless rank_1 $ fail_ctxt "higher-rank"
+      traverse_ go_higher_order_ty ctxt
+      (args', ctxt') <- take_ctxt args
+      pure (args', ctxt ++ ctxt')
+    take_ctxt (DFAForalls fvf _ _) =
       case fvf of
         ForallInvis -> fail_forall "nested"
         ForallVis   -> fail_vdq
-    go_ctxt _ args = go_anons args
+    take_ctxt args = pure (args, [])
 
-    go_anons :: DFunArgs -> m ()
-    go_anons (DFAAnon anon args) = do
-      go_ty False anon
-      go_anons args
-    go_anons (DFAForalls fvf _ _) =
+    take_anons :: DFunArgs -> UnravelM [DType]
+    take_anons (DFAAnon anon args) = do
+      go_higher_order_ty anon
+      anons <- take_anons args
+      pure (anon:anons)
+    take_anons (DFAForalls fvf _ _) =
       case fvf of
         ForallInvis -> fail_forall "nested"
         ForallVis   -> fail_vdq
-    go_anons (DFACxt _ _) = fail_ctxt "nested"
-    go_anons DFANil = pure ()
+    take_anons (DFACxt _ _) = fail_ctxt "nested"
+    take_anons DFANil = pure []
 
-    failWith :: String -> m ()
-    failWith thing = fail $ unlines
+    failWith :: MonadError String m => String -> m a
+    failWith thing = throwError $ unlines
       [ "`singletons` does not support " ++ thing
       , "In the type: " ++ pprint (sweeten ty)
       ]
 
-    fail_forall :: String -> m ()
+    fail_forall :: MonadError String m => String -> m a
     fail_forall sort = failWith $ sort ++ " `forall`s"
 
-    fail_vdq :: m ()
+    fail_vdq :: MonadError String m => m a
     fail_vdq = failWith "visible dependent quantification"
 
-    fail_ctxt :: String -> m ()
+    fail_ctxt :: MonadError String m => String -> m a
     fail_ctxt sort = failWith $ sort ++ " contexts"
+
+-- The monad that powers the internals of unravelVanillaDType_either.
+--
+-- * ExceptT String: records the error message upon failure.
+--
+-- * Reader Bool: True if we are in a rank-1 position in a type, False otherwise
+newtype UnravelM a = UnravelM { runUnravelM :: ExceptT String (Reader Bool) a }
+  deriving (Functor, Applicative, Monad, MonadError String, MonadReader Bool)
 
 -- count the number of arguments in a type
 countArgs :: DType -> Int

--- a/tests/SingletonsTestSuite.hs
+++ b/tests/SingletonsTestSuite.hs
@@ -117,6 +117,7 @@ tests =
     , compileAndDumpStdTest "T367"
     , compileAndDumpStdTest "T371"
     , compileAndDumpStdTest "T376"
+    , compileAndDumpStdTest "T401"
     , compileAndDumpStdTest "T402"
     ],
     testCompileAndDumpGroup "Promote"

--- a/tests/compile-and-dump/Singletons/T401.ghc88.template
+++ b/tests/compile-and-dump/Singletons/T401.ghc88.template
@@ -1,0 +1,13 @@
+
+Singletons/T401.hs:0:0: error:
+    `singletons` does not support higher-rank `forall`s
+In the type: (forall a_0 . a_0 -> a_0) -> b_1 -> b_1
+
+  |
+5 | $(singletons [d|
+  |   ^^^^^^^^^^^^^^...
+
+Singletons/T401.hs:0:0: error: Q monad failure
+  |
+5 | $(singletons [d|
+  |   ^^^^^^^^^^^^^^...

--- a/tests/compile-and-dump/Singletons/T401.hs
+++ b/tests/compile-and-dump/Singletons/T401.hs
@@ -1,0 +1,8 @@
+module T401 where
+
+import Data.Singletons.TH
+
+$(singletons [d|
+  f :: (forall a. a -> a) -> b -> b
+  f _ x = x
+  |])

